feathr_project/feathr/client.py

```diff
@@ -568,13 +569,58 @@ def monitor_features(self, settings: MonitoringSettings, execution_configuration
+    # Return related keys(key_column list) or None if cannot find the feature
+    def _get_feature_key(self, feature_name: str):
+        features = []
+        if 'derived_feature_list' in dir(self):
+            features += self.derived_feature_list 
+        if 'anchor_list' in dir(self):
+            for anchor in self.anchor_list:
+                features += anchor.features  
+        for feature in features:
+            if feature.name == feature_name:
+                keys = feature.key
+                return [key.key_column for key in keys] 
+        self.logger.warning(f"Invalid feature name: {feature_name}. Please call FeathrClient.build_features() first in order to materialize the features.")
+        return None
+        
+    # Validation on feature keys:
+    # Features within a set of aggregation or planned to be merged should have same keys
+    # The param "allow_empty_key" shows if empty keys are acceptable 
+    def _valid_materialize_keys(self, features: List[str], allow_empty_key=False):
+        keys = None
+        for feature in features:
+            new_keys = self._get_feature_key(feature)
+            if new_keys is None:
+                self.logger.error(f"Failed to get feature key for feature: {feature}")
+                return False
+            # If only get one key are it's "NOT_NEEDED", it means the feature has an empty key.
+            if len(new_keys) == 1 and new_keys[0] == "NOT_NEEDED" and not allow_empty_key:
+                self.logger.error(f"Empty feature key is not allowed for features: {features}")
+                return False
+            # Sorted keys to make it easier to compare
+            new_keys = sorted(new_keys)
+            if keys is None:
+                keys = copy.deepcopy(new_keys)
+            else:
+                for key, new_key in zip(keys, new_keys):
+                    if key != new_key:
+                        self.logger.error("Inconsistent feature keys.")
+                        return False
+        return True
+    
     def materialize_features(self, settings: MaterializationSettings, execution_configurations: Union[SparkExecutionConfiguration ,Dict[str,str]] = {}, verbose: bool = False):
         """Materialize feature data
 
         Args:
             settings: Feature materialization settings
             execution_configurations: a dict that will be passed to spark job when the job starts up, i.e. the "spark configurations". Note that not all of the configuration will be honored since some of the configurations are managed by the Spark platform, such as Databricks or Azure Synapse. Refer to the [spark documentation](https://spark.apache.org/docs/latest/configuration.html) for a complete list of spark configurations.
         """
+        feature_list = settings.feature_names
+        if len(feature_list) > 0 and not self._valid_materialize_keys(feature_list):
+            raise RuntimeError(f"Invalid materialization feature keys: {feature_list}")
```

Comment:
The error message should be more clear. Might be something like:

Invalid materialization feature keys: {feature_list}, since they have different keys. Currently Feathr only supports materializing features of the same keys.

Reply:
Changed.